luaL_ref(L, lua.LUA_REGISTRYINDEX) fails

Hi, I'm trying to use the Lua's reference system as described in https://www.lua.org/pil/27.3.2.html, but luaL_ref throws an assertion error:
```
AssertionError: false == true
    at Object.luaH_getint (webpack-internal:///11:153:5)
    at Object.lua_rawgeti (webpack-internal:///63:478:33)
    at Object.luaL_ref (webpack-internal:///5:626:13)
```

Chrome's debugger screenshot:
![image](https://user-images.githubusercontent.com/1613158/30039651-f00d0d7a-91dc-11e7-9627-d007803b7eac.png)

Here's a simplified test case:

```javascript
"use strict";

const test       = require('tape');

global.WEB = false;

const lauxlib    = require("../../src/lauxlib.js");
const lua        = require('../../src/lua.js');

test('luaL_ref, lua_rawgeti, luaL_unref, LUA_REGISTRYINDEX', function( t ) {
    let L;

    t.plan(2)

    t.doesNotThrow(function () {
        L = lauxlib.luaL_newstate();
        lua.lua_pushstring(L, lua.to_luastring("hello references!"))

        let r = lauxlib.luaL_ref(L, lua.LUA_REGISTRYINDEX) // pops a value, stores it and returns a reference
        lua.lua_rawgeti(L, lua.LUA_REGISTRYINDEX, r) // pushes a value associated with the reference
        lauxlib.luaL_unref(L, lua.LUA_REGISTRYINDEX, r) // releases the reference

    }, "JS Lua program ran without error")

    t.strictEqual(
        lua.lua_tojsstring(L, -1),
        "hello references!",
        "top is correct"
    )
});
```

--- a/README.md
+++ b/README.md
@@ -40,6 +40,12 @@ In the browser `require` and `package.loadlib` try to find a file by making sync
     - `os.execute`
 - [Weak tables](http://www.lua.org/manual/5.3/manual.html#2.5.2)
 
+
+### _Differences_ from C API
+
+- `lua_tointegerx` and `lua_tonumberx` do not have out-parameters indicating conversion success. Instead, ``false`` is returned when conversion fails.
+
+
 ## Extensions
 
 ### `dv = lua_todataview(L, idx)`

--- a/src/lapi.js
+++ b/src/lapi.js
@@ -708,11 +708,16 @@ const lua_tocfunction = function(L, idx) {
 };
 
 const lua_tointeger = function(L, idx) {
+    let n = lua_tointegerx(L, idx);
+    return n === false ? 0 : n;
+};
+
+const lua_tointegerx = function(L, idx) {
     return lvm.tointeger(index2addr(L, idx));
 };
 
 const lua_tonumber = function(L, idx) {
-    let n = lvm.tonumber(index2addr(L, idx));
+    let n = lua_tonumberx(L, idx);
     return n === false ? 0 : n;
 };
 
@@ -812,10 +817,6 @@ const lua_stringtonumber = function(L, s) {
         return s.length+1;
     }
     return 0;
-};
-
-const lua_tointegerx = function(L, idx) {
-    return lvm.tointeger(index2addr(L, idx));
 };
 
 const f_call = function(L, ud) {

--- a/src/lapi.js
+++ b/src/lapi.js
@@ -437,7 +437,7 @@ const lua_rawset = function(L, idx) {
 };
 
 const lua_rawseti = function(L, idx, n) {
-    assert(2 < L.top - L.ci.funcOff, "not enough elements in the stack");
+    assert(1 < L.top - L.ci.funcOff, "not enough elements in the stack");
     let o = index2addr(L, idx);
     assert(o.ttistable(), "table expected");
     ltable.luaH_setint(o.value, n, L.stack[L.top - 1]);

--- a/src/lauxlib.js
+++ b/src/lauxlib.js
@@ -343,7 +343,7 @@ const luaL_optnumber = function(L, arg, def) {
 };
 
 const luaL_checkinteger = function(L, arg) {
-    let d = lua.lua_tointeger(L, arg);
+    let d = lua.lua_tointegerx(L, arg);
     if (d === false)
         interror(L, arg);
     return d;

--- a/tests/lauxlib.js
+++ b/tests/lauxlib.js
@@ -1,0 +1,30 @@
+"use strict";
+
+const test       = require('tape');
+
+global.WEB = false;
+
+const lua        = require('../src/lua.js');
+const lauxlib    = require("../src/lauxlib.js");
+
+test('luaL_ref, lua_rawgeti, luaL_unref, LUA_REGISTRYINDEX', function (t) {
+    let L;
+
+    t.plan(2);
+
+    t.doesNotThrow(function () {
+        L = lauxlib.luaL_newstate();
+        lua.lua_pushstring(L, lua.to_luastring("hello references!"));
+
+        let r = lauxlib.luaL_ref(L, lua.LUA_REGISTRYINDEX); // pops a value, stores it and returns a reference
+        lua.lua_rawgeti(L, lua.LUA_REGISTRYINDEX, r); // pushes a value associated with the reference
+        lauxlib.luaL_unref(L, lua.LUA_REGISTRYINDEX, r); // releases the reference
+
+    }, "JS Lua program ran without error");
+
+    t.strictEqual(
+        lua.lua_tojsstring(L, -1),
+        "hello references!",
+        "top is correct"
+    );
+});

